### PR TITLE
Use other methods than FFT to compute the PSD

### PR DESCRIPTION
**Pull request by jbschiratti**
This PR aims at addressing #40 : the estimation of the PSD from the FFT of the data can sometimes be quite noisy. We propose to address that by allowing the PSD to be computed using Welch method (in case of "sufficiently large" windows, averaging should smooth the PSD a bit...), FFT or Multitaper.

The example below shows that using Welch method instead of the FFT can actually improve cross-validation classification scores (using ratios of band power as features).

Example:

```
import mne
import numpy as np
from mne.datasets import sample
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_val_score, KFold
from sklearn.pipeline import Pipeline
from sklearn.base import clone
from sklearn.preprocessing import StandardScaler

from mne_features.feature_extraction import extract_features

###############################################################################
# Let us import the data using MNE-Python and epoch it:

data_path = sample.data_path()
raw_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw.fif'
event_fname = data_path + '/MEG/sample/sample_audvis_filt-0-40_raw-eve.fif'
tmin, tmax = -0.2, 0.5
event_id = dict(aud_l=1, vis_l=3)

# Setup for reading the raw data
raw = mne.io.read_raw_fif(raw_fname, preload=True)
raw.filter(.5, None, fir_design='firwin')
events = mne.read_events(event_fname)
picks = mne.pick_types(raw.info, meg='grad', eeg=False)

# Read epochs
epochs = mne.Epochs(raw, events, event_id, tmin, tmax, picks=picks, proj=True,
                    baseline=None, preload=True)
labels = epochs.events[:, -1]

# get MEG and EEG data
data = epochs.get_data()

###############################################################################
# Prepare for the classification task:
pipe = Pipeline([('scaler', StandardScaler()),
                 ('lr', LogisticRegression(random_state=42, solver='lbfgs'))])
y = labels

###############################################################################
# Cross-validation score using different methods to compute the PSD of the
# signal (FFT, Welch or Multitaper).

kf = KFold(n_splits=3, random_state=42)
selected_funcs = {'pow_freq_bands'}
selected_freq_bands = np.array([0.5, 4., 8., 13., 30., 70.])

for psd_method in ['fft', 'welch', 'multitaper']:

    X = extract_features(
        data, raw.info['sfreq'], selected_funcs,
        funcs_params={'pow_freq_bands__ratios': 'all',
                      'pow_freq_bands__freq_bands': selected_freq_bands,
                      'pow_freq_bands__psd_method': psd_method})
    scores = cross_val_score(clone(pipe), X, y, cv=kf)

    ###########################################################################
    # Print the cross-validation score:

    print('Cross-validation accuracy score with %s method = %1.3f '
          '(+/- %1.5f)' % (psd_method, np.mean(scores), np.std(scores)))
```
Outputs: 
```
Cross-validation accuracy score with fft method = 0.848 (+/- 0.02690)
Cross-validation accuracy score with welch method = 0.910 (+/- 0.02615)
Cross-validation accuracy score with multitaper method = 0.869 (+/- 0.05505)
```

### CODE REVIEW

**Comment by agramfort** [comment]
can you run it with 10 splits for std is more meaningful ?

File: examples/plot_psd_slope_estimation.py [hunk 1]
```diff
@@ -76,15 +75,15 @@
...
 
-# Plot the PSD together with the ``b / (f ** a)`` curve (estimated decay of
-# the PSD with frequency).
+# Plot the PSD together with the ``b + a * f`` straight line (estimated decay
+# of the PSD with frequency in the log10-log10 scale).
 plt.figure()
 plt.semilogx(freqs, np.log10(psd), '-b', lw=2, label='PSD')
 plt.semilogx(freqs, intercept + slope * np.log10(freqs),
-             '-r', lw=2, label='b / (f ** a)')
+             '-r', lw=2, label='b + a * f')
```
Comment (agramfort) [inline examples/plot_psd_slope_estimation.py]: are you sure?

File: mne_features/univariate.py [hunk 2]
```diff
@@ -609,7 +613,7 @@ def compute_pow_freq_bands(sfreq, data, freq_bands=np.array([0.5, 4., 8., 13.,
     n_channels = data.shape[0]
     fb = _freq_bands_helper(sfreq, freq_bands)
     n_freq_bands = fb.shape[0]
-    psd, freqs = power_spectrum(sfreq, data, return_db=False)
+    psd, freqs = power_spectrum(sfreq, data, method=psd_method)
```
Comment (agramfort) [inline mne_features/univariate.py]: you don't expose any of the parameters of Welch or mutlitaper? like the window length and overlap for welch

Comment (jbschiratti) [reply mne_features/univariate.py]: I hesitated to expose them but you're right. I should have !

**Comment by jbschiratti** [comment]
Using `cv = ShuffleSplit(n_splits=10, test_size=0.5, random_state=42)` in `cross_val_score`, I get:
```
Cross-validation accuracy score with fft method = 0.790 (+/- 0.04787)
Cross-validation accuracy score with welch method = 0.867 (+/- 0.01942)
Cross-validation accuracy score with multitaper method = 0.819 (+/- 0.04323)
```
Looks like it's better with Welch anyway.

File: examples/plot_psd_slope_estimation.py (hunk 1 above)
Comment (jbschiratti) [reply examples/plot_psd_slope_estimation.py]: Data to be plotted on the y axis (no log applied on this axis with `semilogx`) is `intercept + slope * np.log10(freqs)`. That's why I changed the legend to `b + a * f`.

Comment (agramfort) [reply examples/plot_psd_slope_estimation.py]: ok

File: mne_features/bivariate.py [hunk 3]
```diff
@@ -347,7 +354,11 @@ def compute_spect_corr(sfreq, data, db=False, with_eigenvalues=True,
            134445/4803/seizure-detection.pdf
     """
     n_channels = data.shape[0]
-    ps, _ = power_spectrum(sfreq, data, return_db=db)
+    if psd_params is not None:
+        psd_params.update({'psd_method': psd_method})
+        ps, _ = power_spectrum(sfreq, data, **psd_params)
+    else:
+        ps, _ = power_spectrum(sfreq, data, psd_method=psd_method)
```
Comment (agramfort) [inline mne_features/bivariate.py]: avoid the if and just do:

`ps, _ = power_spectrum(sfreq, data, psd_method=psd_method, **psd_params)`

File: mne_features/univariate.py [hunk 4]
```diff
@@ -617,7 +628,11 @@ def compute_pow_freq_bands(sfreq, data, freq_bands=np.array([0.5, 4., 8., 13.,
         _freq_bands = np.asarray(freq_bands)
     fb = _freq_bands_helper(sfreq, _freq_bands)
     n_freq_bands = fb.shape[0]
-    psd, freqs = power_spectrum(sfreq, data, return_db=False)
+    if psd_params is not None:
+        psd_params.update({'psd_method': psd_method})
+        psd, freqs = power_spectrum(sfreq, data, **psd_params)
+    else:
+        psd, freqs = power_spectrum(sfreq, data, psd_method=psd_method)
```
Comment (agramfort) [inline mne_features/univariate.py]: same here

File: mne_features/univariate.py [hunk 5]
```diff
@@ -743,7 +785,11 @@ def compute_hjorth_complexity_spect(sfreq, data, normalize=False):
            studies on the prediction of epileptic seizures. Journal of
            Neuroscience Methods, 200(2), 257-271.
     """
-    psd, freqs = power_spectrum(sfreq, data)
+    if psd_params is not None:
+        psd_params.update({'psd_method': psd_method})
+        psd, freqs = power_spectrum(sfreq, data, **psd_params)
+    else:
+        psd, freqs = power_spectrum(sfreq, data, psd_method=psd_method)
```
Comment (agramfort) [inline mne_features/univariate.py]: idem

**Comment by agramfort** [comment]
merge if you're done

does it mean we should rerun the results from iccassp paper ? as I suspect you changed the defaults results with this PR.

**Comment by jbschiratti** [comment]
You're right, Welch method is used as default. Still, setting `psd_method='fft'` will ensure the same behavior (and results) as before.

**Comment by agramfort** [comment]
ok
